### PR TITLE
SINGA-474 ELU operator

### PR DESCRIPTION
**Pull request by ShichengChen**
Implement ONNX Operators following https://github.com/onnx/onnx/blob/master/docs/Operators.md

### CODE REVIEW

File: python/singa/autograd.py [hunk 1]
```diff
@@ -358,6 +358,37 @@ def relu(x):
...
+        super(Elu, self).__init__()
+        self.alpha=alpha
+
+    def forward(self, x):
+        #f(x) = alpha * (exp(x) - 1.) for x < 0, f(x) = x for x >= 0
+        if training:
+            self.input = x
+        x1 = singa.LTFloat(x, 0.0)
+        x1 = singa.__mul__(x, x1)
+        x1 = singa.MultFloat(singa.SubFloat(singa.Exp(x1),self.alpha),1)
```
Comment (nudles) [inline python/singa/autograd.py]: this line is computing (exp(x)-alpha) * 1?

File: test/python/test_operation.py [hunk 2]
```diff
@@ -323,5 +323,30 @@ def test_LeakyRelu(self):
...
+        #f(x) = alpha * (exp(x) - 1.) for x < 0, f(x) = x for x >= 0
+        x = np.array([-0.9, -0.3, -0.1, 0.1, 0.5, 0.9]).reshape(3, 2).astype(np.float32)
+        y = np.clip(x, 0, np.inf) + (np.exp(np.clip(x, -np.inf, 0)) - 1) * 1.0
+        x = tensor.from_numpy(x)
+        x.to_device(cpu_dev)
+
+        result = autograd.elu(x)
+        dx = result.creator.backward(x.data)
+
+        np.testing.assert_array_almost_equal(tensor.to_numpy(result), y, decimal=5)
```
Comment (nudles) [inline test/python/test_operation.py]: only forward is tested..
how about the backward?

File: python/singa/autograd.py (hunk 1 above)
Comment (ShichengChen) [reply python/singa/autograd.py]: yes, I just remove *1

File: test/python/test_operation.py (hunk 2 above)
Comment (ShichengChen) [reply test/python/test_operation.py]: yes, I added backward test

File: python/singa/autograd.py [hunk 3]
```diff
@@ -358,6 +358,48 @@ def relu(x):
...
+        """
+        #f(x) = alpha * (exp(x) - 1.) for x < 0, f(x) = x for x >= 0
+        if training:
+            self.input = x
+        x1 = singa.LTFloat(x, 0.0)
+        x1 = singa.__mul__(x, x1)
+        x1 = singa.SubFloat(singa.Exp(x1),self.alpha)
+        x2 = singa.ReLU(x)
+        x1 = singa.__add__(x1, x2)
+        return x1
```
Comment (joddiy) [inline python/singa/autograd.py]: It's a little weird to call ReLU here, can we implement it directly?

File: python/singa/autograd.py [hunk 4]
```diff
@@ -358,6 +358,48 @@ def relu(x):
...
+
+    def backward(self, dy):
+        """
+        Args:
+            dy (CTensor): data for the dL / dy, L is the loss
+        Returns:
+            a tuple for dx
+        """
+        dx1mask = singa.LTFloat(self.input, 0.0)
+        dx1 = singa.MultFloat(singa.Exp(self.input), self.alpha)
```
Comment (joddiy) [inline python/singa/autograd.py]: is this formula correct?

File: test/python/test_operation.py [hunk 5]
```diff
@@ -322,6 +336,34 @@ def test_LeakyRelu(self):
         np.testing.assert_array_almost_equal(tensor.to_numpy(result), XT)
         self.check_shape(dx.shape(), (3, 2))
 
+    def test_Elu_cpu(self):
+        #f(x) = alpha * (exp(x) - 1.) for x < 0, f(x) = x for x >= 0
+        x = np.array([-0.9, -0.3, -0.1, 0.1, 0.5, 0.9]).reshape(3, 2).astype(np.float32)/10
+        lossf = lambda x : np.sum(np.clip(x, 0, np.inf) + (np.exp(np.clip(x, -np.inf, 0)) - 1) * 1.0)
```
Comment (joddiy) [inline test/python/test_operation.py]: lambda x : alpha * (np.exp(x) -1) if x < 0 else x

File: python/singa/autograd.py [hunk 6]
```diff
@@ -358,6 +358,48 @@ def relu(x):
...
+        dx1 = singa.__mul__(dx1mask, dx1)
+
+        dx2mask = singa.GTFloat(self.input, 0.0)
```
Comment (joddiy) [inline python/singa/autograd.py]: should be GEFloat

File: python/singa/autograd.py [hunk 7]
```diff
@@ -358,6 +358,48 @@ def relu(x):
...
+            x (CTensor): matrix
+        Returns:
+            a CTensor for the result
+        """
+        #f(x) = alpha * (exp(x) - 1.) for x < 0, f(x) = x for x >= 0
+        if training:
+            self.input = x
+        x1 = singa.LTFloat(x, 0.0)
+        x1 = singa.__mul__(x, x1)
+        x1 = singa.SubFloat(singa.Exp(x1),self.alpha)
```
Comment (joddiy) [inline python/singa/autograd.py]: I think this formula doesn't correspond to alpha * (exp(x) - 1.)

Comment (ShichengChen) [reply python/singa/autograd.py]: sorry for my fault, I just modified

**Pull request closed without merging**